// zigbee-herdsman-converters/lib/ota/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.setDataDir = exports.gmmts = exports.jethome = exports.zigbeeOTA = exports.ubisys = exports.tradfri = exports.securifi = exports.lixee = exports.salus = exports.ledvance = exports.inovelli = void 0;
const inovelli = __importStar(require("./inovelli"));
exports.inovelli = inovelli;
const ledvance = __importStar(require("./ledvance"));
exports.ledvance = ledvance;
const salus = __importStar(require("./salus"));
exports.salus = salus;
const lixee = __importStar(require("./lixee"));
exports.lixee = lixee;
const securifi = __importStar(require("./securifi"));
exports.securifi = securifi;
const tradfri = __importStar(require("./tradfri"));
exports.tradfri = tradfri;
const ubisys = __importStar(require("./ubisys"));
exports.ubisys = ubisys;
const common = __importStar(require("./common"));
const zigbeeOTA = __importStar(require("./zigbeeOTA"));
exports.zigbeeOTA = zigbeeOTA;
const jethome = __importStar(require("./jethome"));
exports.jethome = jethome;
const gmmts = __importStar(require("./gmmts"));
exports.gmmts = gmmts;
const { setDataDir } = common;
exports.setDataDir = setDataDir;
