// zigbee-herdsman/dist/utils/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Utils = exports.BackupUtils = exports.RealpathSync = exports.EqualsPartial = exports.Waitress = exports.Queue = exports.Wait = void 0;
const BackupUtils = __importStar(require("./backup"));
exports.BackupUtils = BackupUtils;
const equalsPartial_1 = __importDefault(require("./equalsPartial"));
exports.EqualsPartial = equalsPartial_1.default;
const queue_1 = __importDefault(require("./queue"));
exports.Queue = queue_1.default;
const realpathSync_1 = __importDefault(require("./realpathSync"));
exports.RealpathSync = realpathSync_1.default;
const Utils = __importStar(require("./utils"));
exports.Utils = Utils;
const wait_1 = __importDefault(require("./wait"));
exports.Wait = wait_1.default;
const waitress_1 = __importDefault(require("./waitress"));
exports.Waitress = waitress_1.default;
