// eslint-plugin-perfectionist/dist/rules/sort-vue-attributes.js
'use strict'
const path = require('node:path')
const createEslintRule = require('../utils/create-eslint-rule.js')
const getGroupNumber = require('../utils/get-group-number.js')
const getSourceCode = require('../utils/get-source-code.js')
const rangeToDiff = require('../utils/range-to-diff.js')
const isPositive = require('../utils/is-positive.js')
const useGroups = require('../utils/use-groups.js')
const sortNodes = require('../utils/sort-nodes.js')
const makeFixes = require('../utils/make-fixes.js')
const complete = require('../utils/complete.js')
const pairwise = require('../utils/pairwise.js')
const compare = require('../utils/compare.js')
const sortVueAttributes = createEslintRule.createEslintRule({
  name: 'sort-vue-attributes',
  meta: {
    type: 'suggestion',
    docs: {
      description: 'Enforce sorted Vue attributes.',
    },
    fixable: 'code',
    schema: [
      {
        type: 'object',
        properties: {
          type: {
            description: 'Specifies the sorting method.',
            type: 'string',
            enum: ['alphabetical', 'natural', 'line-length'],
          },
          order: {
            description:
              'Determines whether the sorted items should be in ascending or descending order.',
            type: 'string',
            enum: ['asc', 'desc'],
          },
          ignoreCase: {
            description:
              'Controls whether sorting should be case-sensitive or not.',
            type: 'boolean',
          },
          groups: {
            description: 'Specifies the order of the groups.',
            type: 'array',
            items: {
              oneOf: [
                {
                  type: 'string',
                },
                {
                  type: 'array',
                  items: {
                    type: 'string',
                  },
                },
              ],
            },
          },
          customGroups: {
            description: 'Specifies custom groups.',
            type: 'object',
            additionalProperties: {
              oneOf: [
                {
                  type: 'string',
                },
                {
                  type: 'array',
                  items: {
                    type: 'string',
                  },
                },
              ],
            },
          },
        },
        additionalProperties: false,
      },
    ],
    messages: {
      unexpectedVueAttributesOrder:
        'Expected "{{right}}" to come before "{{left}}".',
    },
  },
  defaultOptions: [
    {
      type: 'alphabetical',
      order: 'asc',
      ignoreCase: true,
      groups: [],
      customGroups: {},
    },
  ],
  create: context => {
    let sourceCode = getSourceCode.getSourceCode(context)
    if (path.extname(context.filename) !== '.vue') {
      return {}
    }
    if (!('defineTemplateBodyVisitor' in sourceCode.parserServices)) {
      return {}
    }
    let { defineTemplateBodyVisitor } = sourceCode.parserServices
    return defineTemplateBodyVisitor({
      VStartTag: node => {
        if (node.attributes.length > 1) {
          let options = complete.complete(context.options.at(0), {
            type: 'alphabetical',
            ignoreCase: true,
            customGroups: {},
            order: 'asc',
            groups: [],
          })
          let parts = node.attributes.reduce(
            (accumulator, attribute) => {
              if (
                attribute.key.type === 'VDirectiveKey' &&
                attribute.key.name.rawName === 'bind'
              ) {
                accumulator.push([])
                return accumulator
              }
              let name
              let { getGroup, defineGroup, setCustomGroups } =
                useGroups.useGroups(options.groups)
              if (
                typeof attribute.key.name === 'string' &&
                attribute.key.type !== 'VDirectiveKey'
              ) {
                name = attribute.key.rawName
              } else {
                name = sourceCode.text.slice(...attribute.key.range)
              }
              setCustomGroups(options.customGroups, name)
              if (attribute.value === null) {
                defineGroup('shorthand')
              }
              if (attribute.loc.start.line !== attribute.loc.end.line) {
                defineGroup('multiline')
              }
              accumulator.at(-1).push({
                size: rangeToDiff.rangeToDiff(attribute.range),
                node: attribute,
                group: getGroup(),
                name,
              })
              return accumulator
            },
            [[]],
          )
          for (let nodes of parts) {
            pairwise.pairwise(nodes, (left, right) => {
              let leftNum = getGroupNumber.getGroupNumber(options.groups, left)
              let rightNum = getGroupNumber.getGroupNumber(
                options.groups,
                right,
              )
              if (
                leftNum > rightNum ||
                (leftNum === rightNum &&
                  isPositive.isPositive(compare.compare(left, right, options)))
              ) {
                context.report({
                  messageId: 'unexpectedVueAttributesOrder',
                  data: {
                    left: left.name,
                    right: right.name,
                  },
                  // @ts-ignore
                  node: right.node,
                  fix: fixer => {
                    let grouped = {}
                    for (let currentNode of nodes) {
                      let groupNum = getGroupNumber.getGroupNumber(
                        options.groups,
                        currentNode,
                      )
                      if (!(groupNum in grouped)) {
                        grouped[groupNum] = [currentNode]
                      } else {
                        grouped[groupNum] = sortNodes.sortNodes(
                          [...grouped[groupNum], currentNode],
                          options,
                        )
                      }
                    }
                    let sortedNodes = []
                    for (let group of Object.keys(grouped).sort(
                      (a, b) => Number(a) - Number(b),
                    )) {
                      sortedNodes.push(
                        ...sortNodes.sortNodes(grouped[group], options),
                      )
                    }
                    return makeFixes.makeFixes(
                      fixer,
                      nodes,
                      sortedNodes,
                      sourceCode,
                    )
                  },
                })
              }
            })
          }
        }
      },
    })
  },
})
module.exports = sortVueAttributes
