// eslint-plugin-perfectionist/dist/rules/sort-array-includes.js
'use strict'
const createEslintRule = require('../utils/create-eslint-rule.js')
const getGroupNumber = require('../utils/get-group-number.js')
const getSourceCode = require('../utils/get-source-code.js')
const toSingleLine = require('../utils/to-single-line.js')
const rangeToDiff = require('../utils/range-to-diff.js')
const isPositive = require('../utils/is-positive.js')
const sortNodes = require('../utils/sort-nodes.js')
const makeFixes = require('../utils/make-fixes.js')
const complete = require('../utils/complete.js')
const pairwise = require('../utils/pairwise.js')
const compare = require('../utils/compare.js')
const sortArrayIncludes = createEslintRule.createEslintRule({
  name: 'sort-array-includes',
  meta: {
    type: 'suggestion',
    docs: {
      description: 'Enforce sorted arrays before include method.',
    },
    fixable: 'code',
    schema: [
      {
        type: 'object',
        properties: {
          type: {
            description: 'Specifies the sorting method.',
            type: 'string',
            enum: ['alphabetical', 'natural', 'line-length'],
          },
          order: {
            description:
              'Determines whether the sorted items should be in ascending or descending order.',
            type: 'string',
            enum: ['asc', 'desc'],
          },
          ignoreCase: {
            description:
              'Controls whether sorting should be case-sensitive or not.',
            type: 'boolean',
          },
          groupKind: {
            description: 'Specifies top-level groups.',
            enum: ['mixed', 'literals-first', 'spreads-first'],
            type: 'string',
          },
        },
        additionalProperties: false,
      },
    ],
    messages: {
      unexpectedArrayIncludesOrder:
        'Expected "{{right}}" to come before "{{left}}".',
    },
  },
  defaultOptions: [
    {
      type: 'alphabetical',
      order: 'asc',
      ignoreCase: true,
      groupKind: 'literals-first',
    },
  ],
  create: context => ({
    MemberExpression: node => {
      if (
        (node.object.type === 'ArrayExpression' ||
          node.object.type === 'NewExpression') &&
        node.property.type === 'Identifier' &&
        node.property.name === 'includes'
      ) {
        let elements =
          node.object.type === 'ArrayExpression'
            ? node.object.elements
            : node.object.arguments
        if (elements.length > 1) {
          let options = complete.complete(context.options.at(0), {
            groupKind: 'literals-first',
            type: 'alphabetical',
            ignoreCase: true,
            order: 'asc',
          })
          let sourceCode = getSourceCode.getSourceCode(context)
          let nodes = elements
            .reduce(
              (accumulator, element) => {
                if (element !== null) {
                  let group = 'unknown'
                  if (typeof options.groupKind === 'string') {
                    group =
                      element.type === 'SpreadElement' ? 'spread' : 'literal'
                  }
                  accumulator.at(0).push({
                    name:
                      element.type === 'Literal'
                        ? `${element.value}`
                        : sourceCode.text.slice(...element.range),
                    size: rangeToDiff.rangeToDiff(element.range),
                    type: element.type,
                    node: element,
                    group,
                  })
                }
                return accumulator
              },
              [[], []],
            )
            .flat()
          pairwise.pairwise(nodes, (left, right) => {
            let groupKindOrder = ['unknown']
            if (typeof options.groupKind === 'string') {
              groupKindOrder =
                options.groupKind === 'literals-first'
                  ? ['literal', 'spread']
                  : ['spread', 'literal']
            }
            let leftNum = getGroupNumber.getGroupNumber(groupKindOrder, left)
            let rightNum = getGroupNumber.getGroupNumber(groupKindOrder, right)
            if (
              (options.groupKind !== 'mixed' && leftNum > rightNum) ||
              ((options.groupKind === 'mixed' || leftNum === rightNum) &&
                isPositive.isPositive(compare.compare(left, right, options)))
            ) {
              context.report({
                messageId: 'unexpectedArrayIncludesOrder',
                data: {
                  left: toSingleLine.toSingleLine(left.name),
                  right: toSingleLine.toSingleLine(right.name),
                },
                node: right.node,
                fix: fixer => {
                  let sortedNodes =
                    options.groupKind !== 'mixed'
                      ? groupKindOrder
                          .map(group => nodes.filter(n => n.group === group))
                          .map(groupedNodes =>
                            sortNodes.sortNodes(groupedNodes, options),
                          )
                          .flat()
                      : sortNodes.sortNodes(nodes, options)
                  return makeFixes.makeFixes(
                    fixer,
                    nodes,
                    sortedNodes,
                    sourceCode,
                  )
                },
              })
            }
          })
        }
      }
    },
  }),
})
module.exports = sortArrayIncludes
