// eslint-plugin-perfectionist/dist/rules/sort-variable-declarations.mjs
import { createEslintRule } from '../utils/create-eslint-rule.mjs'
import { getSourceCode } from '../utils/get-source-code.mjs'
import { toSingleLine } from '../utils/to-single-line.mjs'
import { rangeToDiff } from '../utils/range-to-diff.mjs'
import { isPositive } from '../utils/is-positive.mjs'
import { sortNodes } from '../utils/sort-nodes.mjs'
import { makeFixes } from '../utils/make-fixes.mjs'
import { complete } from '../utils/complete.mjs'
import { pairwise } from '../utils/pairwise.mjs'
import { compare } from '../utils/compare.mjs'
const sortVariableDeclarations = createEslintRule({
  name: 'sort-variable-declarations',
  meta: {
    type: 'suggestion',
    docs: {
      description: 'Enforce sorted variable declarations.',
    },
    fixable: 'code',
    schema: [
      {
        type: 'object',
        properties: {
          type: {
            description: 'Specifies the sorting method.',
            type: 'string',
            enum: ['alphabetical', 'natural', 'line-length'],
          },
          order: {
            description:
              'Determines whether the sorted items should be in ascending or descending order.',
            type: 'string',
            enum: ['asc', 'desc'],
          },
          ignoreCase: {
            description:
              'Controls whether sorting should be case-sensitive or not.',
            type: 'boolean',
          },
        },
        additionalProperties: false,
      },
    ],
    messages: {
      unexpectedVariableDeclarationsOrder:
        'Expected "{{right}}" to come before "{{left}}".',
    },
  },
  defaultOptions: [
    {
      type: 'alphabetical',
      order: 'asc',
      ignoreCase: true,
    },
  ],
  create: context => ({
    VariableDeclaration: node => {
      if (node.declarations.length > 1) {
        let options = complete(context.options.at(0), {
          type: 'alphabetical',
          ignoreCase: true,
          order: 'asc',
        })
        let sourceCode = getSourceCode(context)
        let extractDependencies = init => {
          if (!init) {
            return []
          }
          let dependencies = []
          let checkNode = nodeValue => {
            if (nodeValue.type === 'Identifier') {
              dependencies.push(nodeValue.name)
            }
            if (
              'body' in nodeValue &&
              nodeValue.body &&
              !Array.isArray(nodeValue.body)
            ) {
              traverseNode(nodeValue.body)
            }
            if ('left' in nodeValue) {
              traverseNode(nodeValue.left)
            }
            if ('right' in nodeValue) {
              traverseNode(nodeValue.right)
            }
            if ('elements' in nodeValue) {
              nodeValue.elements
                .filter(currentNode => currentNode !== null)
                .forEach(traverseNode)
            } else if ('arguments' in nodeValue) {
              nodeValue.arguments.forEach(traverseNode)
            }
          }
          let traverseNode = nodeValue => {
            checkNode(nodeValue)
          }
          traverseNode(init)
          return dependencies
        }
        let nodes = node.declarations.map(declaration => {
          let name
          if (
            declaration.id.type === 'ArrayPattern' ||
            declaration.id.type === 'ObjectPattern'
          ) {
            name = sourceCode.text.slice(...declaration.id.range)
          } else {
            ;({ name } = declaration.id)
          }
          let dependencies = extractDependencies(declaration.init)
          return {
            size: rangeToDiff(declaration.range),
            node: declaration,
            dependencies,
            name,
          }
        })
        pairwise(nodes, (left, right) => {
          if (isPositive(compare(left, right, options))) {
            context.report({
              messageId: 'unexpectedVariableDeclarationsOrder',
              data: {
                left: toSingleLine(left.name),
                right: toSingleLine(right.name),
              },
              node: right.node,
              fix: fixer =>
                makeFixes(fixer, nodes, sortNodes(nodes, options), sourceCode),
            })
          }
        })
      }
    },
  }),
})
export { sortVariableDeclarations as default }
