// eslint-plugin-perfectionist/dist/rules/sort-maps.mjs
import { createEslintRule } from '../utils/create-eslint-rule.mjs'
import { getSourceCode } from '../utils/get-source-code.mjs'
import { toSingleLine } from '../utils/to-single-line.mjs'
import { rangeToDiff } from '../utils/range-to-diff.mjs'
import { isPositive } from '../utils/is-positive.mjs'
import { sortNodes } from '../utils/sort-nodes.mjs'
import { makeFixes } from '../utils/make-fixes.mjs'
import { complete } from '../utils/complete.mjs'
import { pairwise } from '../utils/pairwise.mjs'
import { compare } from '../utils/compare.mjs'
const sortMaps = createEslintRule({
  name: 'sort-maps',
  meta: {
    type: 'suggestion',
    docs: {
      description: 'Enforce sorted Map elements.',
    },
    fixable: 'code',
    schema: [
      {
        type: 'object',
        properties: {
          type: {
            description: 'Specifies the sorting method.',
            type: 'string',
            enum: ['alphabetical', 'natural', 'line-length'],
          },
          order: {
            description:
              'Determines whether the sorted items should be in ascending or descending order.',
            type: 'string',
            enum: ['asc', 'desc'],
          },
          ignoreCase: {
            description:
              'Controls whether sorting should be case-sensitive or not.',
            type: 'boolean',
          },
        },
        additionalProperties: false,
      },
    ],
    messages: {
      unexpectedMapElementsOrder:
        'Expected "{{right}}" to come before "{{left}}".',
    },
  },
  defaultOptions: [
    {
      type: 'alphabetical',
      order: 'asc',
      ignoreCase: true,
    },
  ],
  create: context => ({
    NewExpression: node => {
      var _a
      if (
        node.callee.type === 'Identifier' &&
        node.callee.name === 'Map' &&
        node.arguments.length &&
        ((_a = node.arguments[0]) == null ? void 0 : _a.type) ===
          'ArrayExpression'
      ) {
        let [{ elements }] = node.arguments
        if (elements.length > 1) {
          let options = complete(context.options.at(0), {
            type: 'alphabetical',
            ignoreCase: true,
            order: 'asc',
          })
          let sourceCode = getSourceCode(context)
          let parts = elements.reduce(
            (accumulator, element) => {
              if (element === null || element.type === 'SpreadElement') {
                accumulator.push([])
              } else {
                accumulator.at(-1).push(element)
              }
              return accumulator
            },
            [[]],
          )
          for (let part of parts) {
            let nodes = part.map(element => {
              let name
              if (element.type === 'ArrayExpression') {
                let [left] = element.elements
                if (!left) {
                  name = `${left}`
                } else if (left.type === 'Literal') {
                  name = left.raw
                } else {
                  name = sourceCode.text.slice(...left.range)
                }
              } else {
                name = sourceCode.text.slice(...element.range)
              }
              return {
                size: rangeToDiff(element.range),
                node: element,
                name,
              }
            })
            pairwise(nodes, (left, right) => {
              if (isPositive(compare(left, right, options))) {
                context.report({
                  messageId: 'unexpectedMapElementsOrder',
                  data: {
                    left: toSingleLine(left.name),
                    right: toSingleLine(right.name),
                  },
                  node: right.node,
                  fix: fixer =>
                    makeFixes(
                      fixer,
                      nodes,
                      sortNodes(nodes, options),
                      sourceCode,
                    ),
                })
              }
            })
          }
        }
      }
    },
  }),
})
export { sortMaps as default }
