// eslint-plugin-perfectionist/dist/rules/sort-interfaces.mjs
import { minimatch } from 'minimatch'
import { createEslintRule } from '../utils/create-eslint-rule.mjs'
import { isMemberOptional } from '../utils/is-member-optional.mjs'
import { getLinesBetween } from '../utils/get-lines-between.mjs'
import { getGroupNumber } from '../utils/get-group-number.mjs'
import { getSourceCode } from '../utils/get-source-code.mjs'
import { toSingleLine } from '../utils/to-single-line.mjs'
import { rangeToDiff } from '../utils/range-to-diff.mjs'
import { isPositive } from '../utils/is-positive.mjs'
import { useGroups } from '../utils/use-groups.mjs'
import { sortNodes } from '../utils/sort-nodes.mjs'
import { makeFixes } from '../utils/make-fixes.mjs'
import { complete } from '../utils/complete.mjs'
import { pairwise } from '../utils/pairwise.mjs'
import { compare } from '../utils/compare.mjs'
const sortInterfaces = createEslintRule({
  name: 'sort-interfaces',
  meta: {
    type: 'suggestion',
    docs: {
      description: 'Enforce sorted interface properties.',
    },
    fixable: 'code',
    schema: [
      {
        type: 'object',
        properties: {
          type: {
            description: 'Specifies the sorting method.',
            type: 'string',
            enum: ['alphabetical', 'natural', 'line-length'],
          },
          order: {
            description:
              'Determines whether the sorted items should be in ascending or descending order.',
            type: 'string',
            enum: ['asc', 'desc'],
          },
          ignoreCase: {
            description:
              'Controls whether sorting should be case-sensitive or not.',
            type: 'boolean',
          },
          ignorePattern: {
            description:
              'Specifies names or patterns for nodes that should be ignored by rule.',
            items: {
              type: 'string',
            },
            type: 'array',
          },
          partitionByNewLine: {
            description:
              'Allows to use spaces to separate the nodes into logical groups.',
            type: 'boolean',
          },
          groupKind: {
            description: 'Specifies the order of optional and required nodes.',
            enum: ['mixed', 'optional-first', 'required-first'],
            type: 'string',
          },
          groups: {
            description: 'Specifies the order of the groups.',
            type: 'array',
            items: {
              oneOf: [
                {
                  type: 'string',
                },
                {
                  type: 'array',
                  items: {
                    type: 'string',
                  },
                },
              ],
            },
          },
          customGroups: {
            description: 'Specifies custom groups.',
            type: 'object',
            additionalProperties: {
              oneOf: [
                {
                  type: 'string',
                },
                {
                  type: 'array',
                  items: {
                    type: 'string',
                  },
                },
              ],
            },
          },
        },
        additionalProperties: false,
      },
    ],
    messages: {
      unexpectedInterfacePropertiesOrder:
        'Expected "{{right}}" to come before "{{left}}".',
    },
  },
  defaultOptions: [
    {
      type: 'alphabetical',
      order: 'asc',
      ignoreCase: true,
      ignorePattern: [],
      partitionByNewLine: false,
      groupKind: 'mixed',
      groups: [],
      customGroups: {},
    },
  ],
  create: context => ({
    TSInterfaceDeclaration: node => {
      if (node.body.body.length > 1) {
        let options = complete(context.options.at(0), {
          partitionByNewLine: false,
          type: 'alphabetical',
          groupKind: 'mixed',
          ignorePattern: [],
          ignoreCase: true,
          customGroups: {},
          order: 'asc',
          groups: [],
        })
        let sourceCode = getSourceCode(context)
        if (
          !options.ignorePattern.some(pattern =>
            minimatch(node.id.name, pattern, {
              nocomment: true,
            }),
          )
        ) {
          let formattedMembers = node.body.body.reduce(
            (accumulator, element) => {
              var _a, _b, _c, _d
              if (element.type === 'TSCallSignatureDeclaration') {
                accumulator.push([])
                return accumulator
              }
              let lastElement =
                (_a = accumulator.at(-1)) == null ? void 0 : _a.at(-1)
              let name
              let { getGroup, defineGroup, setCustomGroups } = useGroups(
                options.groups,
              )
              if (element.type === 'TSPropertySignature') {
                if (element.key.type === 'Identifier') {
                  ;({ name } = element.key)
                } else if (element.key.type === 'Literal') {
                  name = `${element.key.value}`
                } else {
                  let end =
                    ((_b = element.typeAnnotation) == null
                      ? void 0
                      : _b.range.at(0)) ??
                    element.range.at(1) - (element.optional ? '?'.length : 0)
                  name = sourceCode.text.slice(element.range.at(0), end)
                }
              } else if (element.type === 'TSIndexSignature') {
                let endIndex =
                  ((_c = element.typeAnnotation) == null
                    ? void 0
                    : _c.range.at(0)) ?? element.range.at(1)
                name = sourceCode.text.slice(element.range.at(0), endIndex)
              } else {
                let endIndex =
                  ((_d = element.returnType) == null
                    ? void 0
                    : _d.range.at(0)) ?? element.range.at(1)
                name = sourceCode.text.slice(element.range.at(0), endIndex)
              }
              let elementSortingNode = {
                size: rangeToDiff(element.range),
                node: element,
                name,
              }
              if (
                options.partitionByNewLine &&
                lastElement &&
                getLinesBetween(sourceCode, lastElement, elementSortingNode)
              ) {
                accumulator.push([])
              }
              setCustomGroups(options.customGroups, name)
              if (element.loc.start.line !== element.loc.end.line) {
                defineGroup('multiline')
              }
              accumulator.at(-1).push({
                ...elementSortingNode,
                group: getGroup(),
              })
              return accumulator
            },
            [[]],
          )
          let toSorted = nodes => {
            let grouped = {}
            for (let currentNode of nodes) {
              let groupNum = getGroupNumber(options.groups, currentNode)
              if (!(groupNum in grouped)) {
                grouped[groupNum] = [currentNode]
              } else {
                grouped[groupNum] = sortNodes(
                  [...grouped[groupNum], currentNode],
                  options,
                )
              }
            }
            let sortedNodes = []
            for (let group of Object.keys(grouped).sort(
              (a, b) => Number(a) - Number(b),
            )) {
              sortedNodes.push(...sortNodes(grouped[group], options))
            }
            return sortedNodes
          }
          let checkGroupSort = (left, right) => {
            let leftNum = getGroupNumber(options.groups, left)
            let rightNum = getGroupNumber(options.groups, right)
            return (
              leftNum > rightNum ||
              (leftNum === rightNum &&
                isPositive(compare(left, right, options)))
            )
          }
          let { groupKind } = options
          let checkOrder = (members, left, right, iteration) => {
            if (groupKind === 'mixed') {
              return checkGroupSort(left, right)
            }
            let switchIndex = members.findIndex(
              (_, i) =>
                i &&
                isMemberOptional(members[i - 1].node) !==
                  isMemberOptional(members[i].node),
            )
            if (iteration < switchIndex && iteration + 1 !== switchIndex) {
              return checkGroupSort(left, right)
            }
            if (isMemberOptional(left.node) !== isMemberOptional(right.node)) {
              return (
                isMemberOptional(left.node) !== (groupKind === 'optional-first')
              )
            }
            return checkGroupSort(left, right)
          }
          for (let nodes of formattedMembers) {
            pairwise(nodes, (left, right, iteration) => {
              if (checkOrder(nodes, left, right, iteration)) {
                context.report({
                  messageId: 'unexpectedInterfacePropertiesOrder',
                  data: {
                    left: toSingleLine(left.name),
                    right: toSingleLine(right.name),
                  },
                  node: right.node,
                  fix: fixer => {
                    let sortedNodes
                    if (groupKind !== 'mixed') {
                      let optionalNodes = nodes.filter(member =>
                        isMemberOptional(member.node),
                      )
                      let requiredNodes = nodes.filter(
                        member => !isMemberOptional(member.node),
                      )
                      sortedNodes =
                        groupKind === 'optional-first'
                          ? [
                              ...toSorted(optionalNodes),
                              ...toSorted(requiredNodes),
                            ]
                          : [
                              ...toSorted(requiredNodes),
                              ...toSorted(optionalNodes),
                            ]
                    } else {
                      sortedNodes = toSorted(nodes)
                    }
                    return makeFixes(fixer, nodes, sortedNodes, sourceCode)
                  },
                })
              }
            })
          }
        }
      }
    },
  }),
})
export { sortInterfaces as default }
