// zigbee-herdsman-converters/devices/frankever.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const exposes = __importStar(require("../lib/exposes"));
const legacy = __importStar(require("../lib/legacy"));
const e = exposes.presets;
const ea = exposes.access;
const definitions = [
    {
        fingerprint: [
            { modelID: 'TS0601', manufacturerName: '_TZE200_wt9agwf3' },
            { modelID: 'TS0601', manufacturerName: '_TZE200_5uodvhgc' },
            { modelID: 'TS0601', manufacturerName: '_TZE200_1n2zev06' },
        ],
        model: 'FK_V02',
        vendor: 'FrankEver',
        description: 'Zigbee smart water valve',
        fromZigbee: [legacy.fz.frankever_valve],
        toZigbee: [legacy.tz.tuya_switch_state, legacy.tz.frankever_threshold, legacy.tz.frankever_timer],
        exposes: [
            e.switch().setAccess('state', ea.STATE_SET),
            e
                .numeric('threshold', exposes.access.STATE_SET)
                .withValueMin(0)
                .withValueMax(100)
                .withUnit('%')
                .withDescription('Valve open percentage (multiple of 10)'),
            e
                .numeric('timer', exposes.access.STATE_SET)
                .withValueMin(0)
                .withValueMax(600)
                .withUnit('min')
                .withDescription('Countdown timer in minutes'),
        ],
    },
];
exports.default = definitions;
module.exports = definitions;
